src/provider/provider_os_memory.c

```diff
@@ -0,0 +1,299 @@
+        if (os_params->traces) {
+            fprintf(
+                stderr,
+                "NUMA binding mode (%i) not supported for VisibilityShared\n",
+                os_params->numa_mode);
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    void *addr = mmap(NULL, size, protection, flags, -1, 0);
+    if (addr == MAP_FAILED) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_MMAP_FAILED;
+        if (os_params->traces) {
+            perror("mmap() failed");
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    if ((alignment > 0) && ((uintptr_t)addr & (alignment - 1))) {
+        if (os_params->traces) {
+            TLS_last_native_error = UMF_OS_RESULT_ERROR_MMAP_ADDR_NOT_ALIGNED;
+            fprintf(stderr,
+                    "mmap() returned address 0x%lx not aligned to %zu (0x%lx) "
+                    "bytes\n",
+                    (uintptr_t)addr, alignment, alignment);
+        }
+        goto err_unmap;
+    }
+
+    int ret = mbind(addr, size, os_params->numa_mode, os_params->nodemask,
+                    os_params->maxnode, os_params->numa_flags);
+    if (ret) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_MBIND_FAILED;
+        if (os_params->traces) {
+            perror("mbind() failed");
+        }
+        goto err_unmap;
+    }
+
+    if (os_params->traces) {
+        int numa_node = -1;
+        int ret = get_mempolicy(&numa_node, NULL, 0, (void *)addr,
+                                MPOL_F_NODE | MPOL_F_ADDR);
+        if (ret) {
+            fprintf(stderr,
+                    "cannot print assigned NUMA node - get_mempolicy() failed "
+                    "(errno = %i)\n",
+                    errno);
+            perror("get_mempolicy()");
+        } else {
+            printf("alloc(%zu) = 0x%lx, assigned NUMA node = %i\n", size,
+                   (uintptr_t)addr, numa_node);
+        }
+    }
+
+    *resultPtr = addr;
+
+    return UMF_RESULT_SUCCESS;
+
+err_unmap:
+    (void)munmap(addr, size);
+    return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+}
+
+static enum umf_result_t os_free(void *provider, void *ptr, size_t size) {
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    int ret = munmap(ptr, size);
+    if (ret) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_MUNMAP_FAILED;
+        if (os_params->traces) {
+            perror("munmap() failed");
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+void os_get_last_native_error(void *provider, const char **ppMessage,
+                              int32_t *pError) {
+    (void)provider; // unused
+
+    if (ppMessage == NULL || pError == NULL) {
+        assert(0);
+        return;
+    }
+
+    *pError = TLS_last_native_error;
+    *ppMessage = Native_error_str[*pError];
```

Comment:
Perhaps it would useful to extend the string error message with a string returned from strerror (we would also need to store errno in the tls probably)?

Reply:
IMHO the API is not prepared for that. I would have to allocate memory for such string - would a user be responsible for freeing it? How can I pass a value of `errno` to a user?

Reply:
Well, we could just have a fixed-size buffer in TLS that would be big enough to fit all possible strings (I think we had something similar in PMDK). Something like 1024B would probably be enough.

Right now, I would argue that the Native_error_str is not super helpful for things like bind ("binding memory to NUMA node failed") - it just repeats what the error code is saying.

Reply:
Done